### ISSUE TITLE
Fix S2201: FP for string.Intern which has side effect

### ISSUE THREAD

**Issue by andrei-epure-sonarsource** [SonarSource/sonar-dotnet#2382]
### Description

`string.Intern` has a side-effect: it actually interns the string if it’s not interned already. Hence, this should not be marked as a bug by [S2201](https://rules.sonarsource.com/csharp/RSPEC-2201).

### Repro steps

Add `string.Intern("foo");` inside [ReturnValueIgnored.cs](https://github.com/SonarSource/sonar-dotnet/blob/7.12.0.8217/sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/ReturnValueIgnored.cs) and run the `ReturnValueIgnored` UT.

### Expected behavior

Should not raise an issue.

### Actual behavior

Raises an issue `'Use the return value of method 'Intern', which has no side effect.'`

### Related information

* SonarC# Version: 7.12

Originally reported on [community.sonarsource.com](https://community.sonarsource.com/t/string-intern-does-have-a-side-effect-but-sonarqube-says-it-doesnt/7217)

**Comment by christophe-zurn-sonarsource** [comment]
Looking at the [documentation ](https://docs.microsoft.com/en-us/dotnet/api/system.string.intern?view=netframework-4.8) for `string.Intern`, the method does not seem to have any side effect.
Only the string returned by the method will be interned, the string given as parameter will not be have been interned, so its reference might still be different than the interned one.

**Comment by christophe-zurn-sonarsource** [comment]
Scratch that, it does have a side effect:
```
If the string does not exist, a reference to str is added to the intern pool, then that reference is returned.
```
Although I don't see why one would not use the return value that holds the reference to the interned string, since it has a side effect, we can definitely add an exception for this rule.

**Comment by christophe-zurn-sonarsource** [comment]
As discussed, we will keep raising issue on unused `string.Intern` function call, but updated message and description
